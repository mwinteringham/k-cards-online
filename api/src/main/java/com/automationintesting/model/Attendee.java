package com.automationintesting.model;

import com.automationintesting.service.CodeGenerator;
import com.fasterxml.jackson.annotation.JsonProperty;

public class Attendee {

    @JsonProperty
    private String name;
    @JsonProperty
    private String code;

    public Attendee() {}

    public Attendee(String name, String code) {
        this.name = name;
        this.code = code;
    }

    public Attendee(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }

    @Override
    public String toString() {
        return "Attendee{" +
                "name='" + name + '\'' +
                ", code='" + code +
                '}';
    }
}
