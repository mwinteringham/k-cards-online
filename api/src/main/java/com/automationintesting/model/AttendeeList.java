package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class AttendeeList {

    @JsonProperty
    private List<Attendee> attendees;

    public AttendeeList() {
    }

    public AttendeeList(List<Attendee> attendees) {
        this.attendees = attendees;
    }

    public List<Attendee> getAttendees() {
        return attendees;
    }

    public void setAttendees(List<Attendee> attendees) {
        this.attendees = attendees;
    }

    @Override
    public String toString() {
        return "AttendeeList{" +
                "attendees=" + attendees +
                '}';
    }
}
