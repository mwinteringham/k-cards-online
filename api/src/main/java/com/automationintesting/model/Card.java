package com.automationintesting.model;

import com.fasterxml.jackson.annotation.JsonProperty;

public class Card {

    @JsonProperty
    private java.lang.String name;
    @JsonProperty
    private String string;

    public Card() {
    }

    public Card(String name, String string) {
        this.name = name;
        this.string = string;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getCardType() {
        return string;
    }

    public void setCardType(String string) {
        this.string = string;
    }
}
