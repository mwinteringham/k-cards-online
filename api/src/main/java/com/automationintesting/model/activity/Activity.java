package com.automationintesting.model.activity;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

public class Activity {

//    "activity": {
//        "reds": [
//          "Amy Lee",
//          "Barry White",
//          "Jerry Jones"
//        ],
//        "threads": [{
//            "name": "Amy Lee"
//        }, {
//            "name": "Barry White",
//            "subthreads": [{
//                "name": "Stevie"
//            }, {
//                "name": "David"
//            }]
//        }]
//    }

    @JsonProperty
    private List<String> reds;
    @JsonProperty
    private List<ActivityThread> threads;

    public Activity() {
    }

    public Activity(List<String> reds, List<ActivityThread> threads) {
        this.reds = reds;
        this.threads = threads;
    }

    public List<String> getReds() {
        return reds;
    }

    public void setReds(List<String> reds) {
        this.reds = reds;
    }

    public List<ActivityThread> getThreads() {
        return threads;
    }

    public void setThreads(List<ActivityThread> threads) {
        this.threads = threads;
    }

    @Override
    public String toString() {
        return "Activity{" +
                "reds=" + reds +
                ", threads=" + threads +
                '}';
    }
}
