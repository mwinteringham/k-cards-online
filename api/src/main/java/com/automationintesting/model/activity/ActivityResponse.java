package com.automationintesting.model.activity;

import com.fasterxml.jackson.annotation.JsonProperty;

public class ActivityResponse {

//    {
//        "activity": {
//            "reds": [
//              "Amy Lee",
//              "Barry White",
//              "Jerry Jones"
//            ],
//            "threads": [{
//                "name": "Amy Lee"
//            }, {
//                "name": "Barry White",
//                "subthreads": [{
//                    "name": "Stevie"
//                }, {
//                    "name": "David"
//                }]
//            }]
//        }
//    }

    @JsonProperty
    private Activity activity;

    public ActivityResponse() {
    }

    public ActivityResponse(Activity activity) {
        this.activity = activity;
    }

    public Activity getActivity() {
        return activity;
    }

    public void setActivity(Activity activity) {
        this.activity = activity;
    }

    @Override
    public String toString() {
        return "ActivityResponse{" +
                "activity=" + activity.toString() +
                '}';
    }
}
