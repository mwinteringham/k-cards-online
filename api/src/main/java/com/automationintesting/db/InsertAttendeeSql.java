package com.automationintesting.db;

import com.automationintesting.model.Attendee;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class InsertAttendeeSql {

    private final PreparedStatement preparedStatement;

    public InsertAttendeeSql(Connection connection, Attendee attendee, String workshop) throws SQLException {
        final String CREATE_ATTENDEE = "INSERT INTO ATTENDEES (attendeecode, name, workshopcode) VALUES (?, ?, ?)";

        preparedStatement = connection.prepareStatement(CREATE_ATTENDEE);
        preparedStatement.setString(1, attendee.getCode());
        preparedStatement.setString(2, attendee.getName());
        preparedStatement.setString(3, workshop);
    }

    public PreparedStatement getPreparedStatement() {
        return preparedStatement;
    }

}
