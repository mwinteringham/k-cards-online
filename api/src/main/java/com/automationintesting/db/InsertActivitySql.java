package com.automationintesting.db;

import com.automationintesting.model.Card;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

public class InsertActivitySql {

    private final PreparedStatement preparedStatement;

    public InsertActivitySql(Connection connection, Card card, String workshopCode) throws SQLException {
        final String CREATE_ACTIVITY = "INSERT INTO ACTIVITIES (cardtype, name, workshopcode) VALUES (?, ?, ?)";

        preparedStatement = connection.prepareStatement(CREATE_ACTIVITY);
        preparedStatement.setString(1, card.getCardType());
        preparedStatement.setString(2, card.getName());
        preparedStatement.setString(3, workshopCode);
    }

    public PreparedStatement getPreparedStatement() {
        return preparedStatement;
    }
}
