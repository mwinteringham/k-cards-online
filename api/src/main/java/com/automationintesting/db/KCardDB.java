package com.automationintesting.db;

import com.automationintesting.model.Attendee;
import com.automationintesting.model.AttendeeList;
import com.automationintesting.model.Card;
import com.automationintesting.model.activity.Activity;
import org.h2.jdbcx.JdbcDataSource;
import org.springframework.stereotype.Component;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

@Component
public class KCardDB {

    private Connection connection;

    private final String CREATE_WORKSHOPS_TABLE = "CREATE table WORKSHOPS ( code varchar(6), workshopName varchar(255), primary key (code))";
    private final String CREATE_ATTENDEE_TABLE = "CREATE table ATTENDEES ( attendee_id int NOT NULL AUTO_INCREMENT, name varchar(255), workshopcode varchar(6), primary key (attendee_id))";
    private final String CREATE_ACTIVITY_TABLE = "CREATE table ACTIVITIES ( activity_id int NOT NULL AUTO_INCREMENT, cardtype varchar(6), name varchar(255), workshopcode varchar(6), primary key (activity_id))";

    public KCardDB() throws SQLException {
        JdbcDataSource ds = new JdbcDataSource();
        ds.setURL("jdbc:h2:mem:kcard;MODE=MySQL");
        ds.setUser("user");
        ds.setPassword("password");
        connection = ds.getConnection();

        connection.prepareStatement(CREATE_WORKSHOPS_TABLE).executeUpdate();
        connection.prepareStatement(CREATE_ATTENDEE_TABLE).executeUpdate();
        connection.prepareStatement(CREATE_ACTIVITY_TABLE).executeUpdate();
    }

    public Boolean addCode(String code, String workshopName) throws SQLException {
        InsertWorkshopSql insertWorkshopSql = new InsertWorkshopSql(connection, code, workshopName);
        PreparedStatement createPs = insertWorkshopSql.getPreparedStatement();

        return createPs.executeUpdate() > 0;
    }

    public Boolean addAttendee(Attendee attendee, String workshopCode) throws SQLException {
        InsertAttendeeSql insertAttendeeSql = new InsertAttendeeSql(connection, attendee, workshopCode);
        PreparedStatement createPs = insertAttendeeSql.getPreparedStatement();

        return createPs.executeUpdate() > 0;
    }

    public AttendeeList getAttendeesInWorkshop(String workshopCode) throws SQLException {
        List<Attendee> listToReturn = new ArrayList<>();
        String sql = "SELECT * FROM ATTENDEES WHERE workshopcode = '" + workshopCode + "'";

        ResultSet results = connection.prepareStatement(sql).executeQuery();

        while(results.next()){
            listToReturn.add(new Attendee(results.getString("name")));
        }

        return new AttendeeList(listToReturn);
    }

    public Boolean addCardActivity(Card card, String workshopCode) throws SQLException {
        InsertActivitySql insertActivitySql = new InsertActivitySql(connection, card, workshopCode);
        PreparedStatement createPs = insertActivitySql.getPreparedStatement();

        return createPs.executeUpdate() > 0;
    }

    public Activity getWorkshopActivity(String workshopCode) throws SQLException {
        List<String> redList = new ArrayList<>();
        String sql = "SELECT * FROM ACTIVITIES WHERE workshopcode = '" + workshopCode + "'";

        ResultSet results = connection.prepareStatement(sql).executeQuery();

        while(results.next()){
            if(results.getString("cardtype").contentEquals("red")){
                redList.add(results.getString("name"));
            }
        }

        return new Activity(redList, null);
    }
}
