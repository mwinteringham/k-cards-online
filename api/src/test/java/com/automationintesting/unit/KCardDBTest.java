package com.automationintesting.unit;

import com.automationintesting.db.KCardDB;
import com.automationintesting.model.Attendee;
import com.automationintesting.model.AttendeeList;
import com.automationintesting.model.Card;
import com.automationintesting.model.activity.Activity;
import com.automationintesting.model.activity.ActivityResponse;
import org.approvaltests.Approvals;
import org.junit.BeforeClass;
import org.junit.Test;

import java.sql.SQLException;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsEqual.equalTo;

public class KCardDBTest {

    private static KCardDB kCardDB;

    @BeforeClass
    public static void setupDb() throws SQLException {
        kCardDB = new KCardDB();
    }

    @Test
    public void storeWorkshopInDb() throws SQLException {
        String workshopCode = "abcdef";
        String workshopName = "LEWT";

        Boolean storeResult = kCardDB.addCode(workshopCode, workshopName);

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void storeAttendeeInDb() throws SQLException {
        Attendee attendee = new Attendee("Stacey Jones");

        Boolean storeResult = kCardDB.addAttendee(attendee, "abcdef");

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void returnAttendeesList() throws SQLException {
        Attendee attendee = new Attendee("Barry White");
        kCardDB.addAttendee(attendee, "xyzxyz");

        AttendeeList attendeeList = kCardDB.getAttendeesInWorkshop("xyzxyz");

        assertThat(attendeeList.getAttendees().size(), equalTo(1));
    }

    @Test
    public void storeCardActivity() throws SQLException {
        Card card = new Card("Amy Lee", "red");

        Boolean storeResult = kCardDB.addCardActivity(card, "jsdjsd");

        assertThat(storeResult, equalTo(true));
    }

    @Test
    public void returnWorkshopActivity() throws SQLException {
        Card card = new Card("Amy Lee", "red");
        kCardDB.addCardActivity(card, "BEWT");

        Activity activity = kCardDB.getWorkshopActivity("BEWT");

        Approvals.verify(activity);
    }

}
