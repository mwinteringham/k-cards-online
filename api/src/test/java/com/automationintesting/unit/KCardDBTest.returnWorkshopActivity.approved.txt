Activity{reds=[Amy Lee], threads=null}
